src/main/scala/org/apache/spark/sql/pulsar/PulsarHelper.scala

```diff
@@ -207,6 +222,39 @@ private[pulsar] case class PulsarHelper(
     }.toMap)
   }
 
+  def latestOffsets(startingOffset: streaming.Offset,
+                    totalReadLimit: Long): SpecificPulsarOffset = {
+    // implement helper inside PulsarHelper in order to use getTopicPartitions
+    val topicPartitions = getTopicPartitions
+    // add new partitions from PulsarAdmin, set to earliest entry and ledger id based on limit
+    // start a reader, get to the earliest offset for new topic partitions
+    val existingStartOffsets = if (startingOffset != null) {
+      getTopicOffsets(startingOffset.asInstanceOf[org.apache.spark.sql.execution.streaming.Offset])
+    } else {
+      Map[String, MessageId]()
+    }
+    val newTopics = topicPartitions.toSet.diff(existingStartOffsets.keySet)
+    logInfo(s"EXISTING TOPIC PARTITIONS: ${existingStartOffsets.keySet.mkString(",")}\n")
+    logInfo(s"ALL TOPIC PARTITIONS: ${topicPartitions.mkString(",")}\n")
+    val startPartitionOffsets = existingStartOffsets ++ newTopics.map(topicPartition
+      => {
+      logInfo(s"SETTING NEW TOPIC PARTITION: $topicPartition\n")
+      topicPartition -> MessageId.earliest
+    })
+    val offsets = mutable.Map[String, MessageId]()
+    offsets ++= startPartitionOffsets
```

Comment:
Is this redundant?

Reply:
The thing is, we need to return offsets, and we can't modify startPartitionOffsets, which is why I do this here.

Reply:
Any key value pair will be overwritten by the foreach loop below, this is actually no-op. Do I miss something?

Reply:
Ah you're right. This was necessary before the code refactor, but not anymore.